staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1285,12 +1175,42 @@ func (c *cacheWatcher) process(ctx context.Context, initEvents []*watchCacheEven
 			if event.ResourceVersion > resourceVersion {
 				c.sendWatchCacheEvent(event)
 			}
+		case <-bookmarkTimer:
+			bookmarkTimer = c.nextBookmarkTimer(ctx)
+			event := bookmarkEvent()
+			if event == nil {
+				continue
+			}
+			// send BOOKMARK event using non-blocking method
+			select {
+			case c.result <- *event:
+			default:
+			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+func (c *cacheWatcher) nextBookmarkTimer(ctx context.Context) <-chan time.Time {
+	now := c.clock.Now()
+
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return time.After(time.Minute)
+	}
+
+	duration := deadline.Add(-2 * time.Second).Sub(now)
+
+	if duration < 0 {
+		return nil
+	} else if duration > time.Minute {
+		return time.After(time.Minute)
+	}
+
+	return time.After(duration)
```

Comment:
This means we're creating a ticker for every single watch (and also are not stopping it when watch is closing). So this may actually lead to much higher resource usage...